plugins/python/redisgears_python.c

```diff
@@ -1751,6 +1759,509 @@ static PyObject* atomicExit(PyObject *self, PyObject *args){
+    verifyRedisAILoaded();
+    PyDAGRunner *pyDag = (PyDAGRunner *)self;
+    if (!_IsValidDag(pyDag, RedisAI_DAGAddOpsFromString)) {
+        return NULL;
+    }
+    if(PyTuple_Size(args) != 1){
+        PyErr_SetString(GearsError, "Wrong number of args");
+        return NULL;
+    }
+    PyObject* ops = PyTuple_GetItem(args, 0);
+    if(!PyUnicode_Check(ops)){
+        PyErr_SetString(GearsError, "ops argument must be a string");
+        return NULL;
+    }
+    const char* opsStr = PyUnicode_AsUTF8AndSize(ops, NULL);
+    RAI_Error *err;
+    RedisAI_InitError(&err);
+    if (RedisAI_DAGAddOpsFromString(pyDag->dag, opsStr, err) != REDISMODULE_OK) {
+        PyErr_SetString(GearsError, RedisAI_GetError(err));
+        RedisAI_FreeError(err);
+        return NULL;
+    }
+    RedisAI_FreeError(err);
+    Py_INCREF(self);
+    return self;
+}
+
+static void FinishAsyncDAGRun(RAI_OnFinishCtx *onFinishCtx, void *private_data) {
+
+    RedisGearsPy_LOCK
+    PyObject* future = private_data;
+    PyObject* pArgs = PyTuple_New(2);
+    PyTuple_SetItem(pArgs, 0, future);
+
+    if (RedisAI_DAGRunError(onFinishCtx)) {
+        PyObject* pyErr;
+        const RAI_Error* err = RedisAI_DAGGetError(onFinishCtx);
+        const char* errStr = (RedisAI_GetError((RAI_Error*)err));
+        pyErr = PyUnicode_FromStringAndSize(errStr, strlen(errStr));
+
+        PyTuple_SetItem(pArgs, 1, pyErr);
+        PyObject_CallObject(setFutureExceptionFunction, pArgs);
+        goto finish;
+    }
+
+    PyObject* tensorList = PyList_New(0);
+    size_t n_outputs = RedisAI_DAGNumOutputs(onFinishCtx);
+    for(size_t i = 0 ; i < n_outputs ; ++i){
+        PyTensor* pyt = PyObject_New(PyTensor, &PyTensorType);
+        pyt->t = RedisAI_TensorGetShallowCopy((RAI_Tensor*)RedisAI_DAGOutputTensor(onFinishCtx, i));
+        PyList_Append(tensorList, (PyObject*)pyt);
+        GearsPyDecRef(pyt);
+    }
+
+    PyTuple_SetItem(pArgs, 1, tensorList);
+    PyObject* r = PyObject_CallObject(setFutureResultsFunction, pArgs);
+    if(!r){
+        char* err = getPyError();
+        RedisModule_Log(NULL, "warning", "Error happened when releasing execution future, error='%s'", err);
+        RG_FREE(err);
+    }else{
+        GearsPyDecRef(r);
+    }
+
+    finish:
+    RedisAI_DAGFree((RAI_DAGRunCtx *)onFinishCtx);
+    GearsPyDecRef(pArgs);
+    RedisGearsPy_UNLOCK
+}
+
+static PyObject* DAGRun(PyObject *self){
+    verifyRedisAILoaded();
+    PyDAGRunner *pyDag = (PyDAGRunner *)self;
+    if (!_IsValidDag(pyDag, RedisAI_DAGRun)) {
+        return NULL;
+    }
+
+    PyObject* pArgs = PyTuple_New(0);
+    PyObject* future = PyObject_CallObject(createFutureFunction, pArgs);
+    GearsPyDecRef(pArgs);
+
+    RAI_Error *err;
+    RedisAI_InitError(&err);
+    int status = RedisAI_DAGRun(pyDag->dag, FinishAsyncDAGRun, future, err);
+    if (status == REDISMODULE_ERR) {
+        PyErr_SetString(GearsError, RedisAI_GetError(err));
```

Comment:
I think it make sense to do `pyDag->dag = NULL;` here also, and also everywhere when there is an error on DAG operation, WDYT?